client/rest-high-level/src/main/java/org/elasticsearch/client/security/support/expressiondsl/expressions/CompositeExpressionBuilder.java

```diff
@@ -0,0 +1,88 @@
+import org.elasticsearch.client.security.support.expressiondsl.fields.DnFieldExpression;
+import org.elasticsearch.client.security.support.expressiondsl.fields.FieldExpressionBuilder;
+import org.elasticsearch.client.security.support.expressiondsl.fields.GroupsFieldExpression;
+import org.elasticsearch.client.security.support.expressiondsl.fields.MetadataFieldExpression;
+import org.elasticsearch.client.security.support.expressiondsl.fields.UsernameFieldExpression;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Builder for composite role mapper expressions.<br>
+ * Example usage:
+ * <pre>
+ * {@code
+ * final AllExpression allExpression = CompositeExpressionBuilder.builder(AllExpression.class)
+                .addExpression(CompositeExpressionBuilder.builder(AnyExpression.class)
+                        .addExpression(FieldExpressionBuilder.builder(DnFieldExpression.class)
+                                .addValue("*,ou=admin,dc=example,dc=com")
+                                .build())
+                        .addExpression(FieldExpressionBuilder.builder(UsernameFieldExpression.class)
+                                .addValue("es-admin").addValue("es-system")
+                                .build())
+                        .build())
+                .addExpression(FieldExpressionBuilder.builder(GroupsFieldExpression.class)
+                        .addValue("cn=people,dc=example,dc=com")
+                        .build())
+                .addExpression(new ExceptExpression(FieldExpressionBuilder.builder(MetadataFieldExpression.class)
+                                            .withKey("metadata.terminated_date")
+                                            .addValue(new Date())
+                                            .build()))
+                .build();
+ * }
+ * </pre>
+ */
+public final class CompositeExpressionBuilder<T extends CompositeRoleMapperExpressionBase> {
+    private Constructor<T> ctor;
+    private List<RoleMapperExpression> elements = new ArrayList<>();
+
+    private CompositeExpressionBuilder(Class<T> clazz) {
+        try {
+            this.ctor = clazz.getConstructor(RoleMapperExpression[].class);
+        } catch (NoSuchMethodException | SecurityException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static <T extends CompositeRoleMapperExpressionBase> CompositeExpressionBuilder<T> builder(Class<T> clazz) {
+        return new CompositeExpressionBuilder<>(clazz);
+    }
+
+    public CompositeExpressionBuilder<T> addExpression(final RoleMapperExpression expression) {
+        assert expression != null : "expression cannot be null";
+        elements.add(expression);
+        return this;
+    }
+
+    public T build() {
+        try {
+            return ctor.newInstance(new Object[] { elements.toArray(new RoleMapperExpression[0]) });
```

Comment:
I don't think we should be using reflection for this.
There's only 2 classes to handle, we can do it more simply and safely with a `Function<RoleMapperExpression[], T>`